File: typescript/src/streaming.ts
/**
 * Refresh sampling as it actually happens: one arrival at a time.
 *
 * Refresh-time sampling is a streaming algorithm wearing a batch costume. A refresh is
 * detected the instant the last required instrument posts — there is nothing to look
 * ahead to, and nothing to wait for beyond the arrival itself. So the live engine is not
 * an approximation of the batch function here; it is the same computation, driven by the
 * feed instead of by a loop.
 *
 * {@link StreamingRefreshSampler.observe} returns the refresh row if that arrival
 * completed one, and `null` otherwise. That single return value is the whole API: the
 * moment a refresh is emitted is the moment it became true.
 *
 * The batch function and this engine are asserted to produce identical rows for
 * identical input, because the day they diverge is the day your research and your
 * production sampler stop describing the same series.
 */

import {
  BEFORE_TIME,
  type NormalisedRow,
  type PartitionSummary,
  type RefreshRow,
  buildRow,
  formatTimestampMs,
  normaliseRow,
  requireInstruments,
  select,
} from "./core.ts";

/** Emit refresh cross-sections as arrivals complete them. */
export class StreamingRefreshSampler {
  readonly #instruments: string[];
  readonly #allowed: Set<string>;
  readonly #maxStalenessMs: number;

  #partition: string | null = null;
  #cursor = BEFORE_TIME;
  #sequence = 0;
  #lastAvailableMs: number | null = null;
  #latestByRecord = new Map<string, NormalisedRow>();
  #histories = new Map<string, NormalisedRow>();
  #eventOwners = new Map<string, string>();
  #pending = new Map<string, NormalisedRow[]>();
  #discardedTotal = 0;
  #accepted = 0;
  #stale = 0;
  #inputUpdates = 0;

  /**
   * @param requiredInstruments The names that must all refresh. At least two.
   * @param maxStalenessMs Age past which a complete row is flagged `stale`.
   *
   * One sampler handles one partition at a time. Feeding it two partitions throws,
   * rather than silently interleaving two unrelated sessions into one series.
   */
  constructor(requiredInstruments: Iterable<unknown>, maxStalenessMs: number) {
    if (
      typeof maxStalenessMs !== "number" ||
      !Number.isInteger(maxStalenessMs) ||
      maxStalenessMs < 0
    ) {
      throw new Error("max_staleness_ms must be a non-negative integer");
    }
    this.#instruments = requireInstruments(requiredInstruments);
    this.#allowed = new Set(this.#instruments);
    this.#maxStalenessMs = maxStalenessMs;
    for (const instrument of this.#instruments) this.#pending.set(instrument, []);
  }

  // --- ingestion ---------------------------------------------------------- //
  /** Ingest one arrival. Returns a refresh row if this arrival completed one. */
  observe(observation: unknown): RefreshRow | null {
    // Per-row validation only. The cross-row revision rules are enforced by
    // #checkHistory below, which carries state the batch validator cannot see.
    const row = normaliseRow(observation, this.#allowed);

    if (this.#partition === null) this.#partition = row.partition;
    else if (row.partition !== this.#partition) {
      throw new Error(
        `one sampler handles one partition; got ${JSON.stringify(row.partition)} ` +
          `after ${JSON.stringify(this.#partition)}`,
      );
    }

    if (this.#lastAvailableMs !== null && row.availableMs < this.#lastAvailableMs) {
      throw new Error("observations must arrive in non-decreasing available_at order");
    }
    this.#checkHistory(row);

    this.#lastAvailableMs = row.availableMs;
    this.#inputUpdates += 1;
    this.#latestByRecord.set(row.record_id, row);
    this.#pending.get(row.instrument)!.push(row);

    for (const queue of this.#pending.values()) {
      if (queue.length === 0) return null;
    }
    return this.#emit(row.availableMs);
  }

  /** Ingest many arrivals, returning every refresh row they completed. */
  observeMany(observations: Iterable<unknown>): RefreshRow[] {
    const emitted: RefreshRow[] = [];
    for (const observation of observations) {
      const row = this.observe(observation);
      if (row !== null) emitted.push(row);
    }
    return emitted;
  }

  #checkHistory(row: NormalisedRow): void {
    const eventKey = `${row.instrument}|${row.eventMs}`;
    const owner = this.#eventOwners.get(eventKey);
    if (owner === undefined) this.#eventOwners.set(eventKey, row.record_id);
    else if (owner !== row.record_id) {
      throw new Error("ambiguous duplicate event_time across distinct record_id values");
    }

    const previous = this.#histories.get(row.record_id);
    if (previous === undefined) {
      if (row.revision !== 0) throw new Error("the first revision of a record must be zero");
    } else {
      if (row.instrument !== previous.instrument) {
        throw new Error("a correction cannot change instrument");
      }
      if (row.revision !== previous.revision + 1) {
        throw new Error("record revisions must be contiguous");
      }
      if (row.availableMs <= previous.availableMs) {
        throw new Error("corrections must become available strictly later");
      }
    }
    this.#histories.set(row.record_id, row);
  }

  #emit(refreshMs: number): RefreshRow {
    const selected = select(this.#latestByRecord, this.#instruments);
    const controllers = [...this.#pending.entries()]
      .filter(([, queue]) => queue[0]!.availableMs === refreshMs)
      .map(([instrument]) => instrument)
      .sort();

    const kept = new Set(
      Object.values(selected)
        .filter((row) => row.availableMs > this.#cursor && row.availableMs <= refreshMs)
        .map((row) => `${row.record_id}|${row.revision}`),
    );
    let intervalCount = 0;
    const arrivalsByInstrument: Record<string, number> = {};
    for (const instrument of this.#instruments) {
      const count = this.#pending.get(instrument)!.length;
      arrivalsByInstrument[instrument] = count;
      intervalCount += count;
    }

    const discarded = intervalCount - kept.size;
    this.#discardedTotal += discarded;
    this.#sequence += 1;

    const emitted = buildRow(
      this.#partition!,
      this.#sequence,
      this.#cursor,
      refreshMs,
      controllers,
      selected,
      this.#instruments,
      arrivalsByInstrument,
      discarded,
      this.#maxStalenessMs,
    );
    if (emitted.status === "stale") this.#stale += 1;
    else this.#accepted += 1;

    this.#cursor = refreshMs;
    // Every queue is non-empty here, so keeping only arrivals after the refresh leaves
    // exactly what has not yet been consumed.
    for (const instrument of this.#instruments) {
      this.#pending.set(
        instrument,
        this.#pending.get(instrument)!.filter((row) => row.availableMs > refreshMs),
      );
    }
    return emitted;
  }

  // --- introspection ------------------------------------------------------ //
  /** The same partition summary the batch function reports. */
  get summary(): PartitionSummary {
    let tail = 0;
    for (const queue of this.#pending.values()) tail += queue.length;
    return {
      partition: this.#partition as string,
      input_updates: this.#inputUpdates,
      refresh_candidates: this.#sequence,
      accepted_rows: this.#accepted,
      stale_rows: this.#stale,
      discarded_updates: this.#discardedTotal,
      unmatched_tail_updates: tail,
      loss_fraction: {
        numerator: this.#discardedTotal + tail,
        denominator: this.#inputUpdates,
      },
    };
  }

  /**
   * Instruments that have not posted since the last refresh.
   *
   * The live version of the controller question: whatever is in this list is what the
   * whole basket is currently waiting on.
   */
  get waitingFor(): string[] {
    return [...this.#pending.entries()]
      .filter(([, queue]) => queue.length === 0)
      .map(([instrument]) => instrument)
      .sort();
  }

  get refreshCount(): number {
    return this.#sequence;
  }

  get lastRefreshAt(): string | null {
    return this.#cursor === BEFORE_TIME ? null : formatTimestampMs(this.#cursor);
  }
}

File: python/src/fintech_refresh_time/streaming.py
"""Refresh sampling as it actually happens: one arrival at a time.

Refresh-time sampling is a streaming algorithm wearing a batch costume. A refresh is
detected the instant the last required instrument posts — there is nothing to look
ahead to, and nothing to wait for beyond the arrival itself. So the live engine is not
an approximation of the batch function here; it is the same computation, driven by the
feed instead of by a loop.

:meth:`StreamingRefreshSampler.observe` returns the refresh row if that arrival
completed one, and ``None`` otherwise. That single return value is the whole API: the
moment a refresh is emitted is the moment it became true.

The batch function and this engine are asserted to produce identical rows for identical
input, because the day they diverge is the day your research and your production
sampler stop describing the same series.
"""

from __future__ import annotations

from typing import Any, Iterable, Mapping

from .core import (
    _BEFORE_TIME,
    _require_instruments,
    _select,
    format_timestamp_ms,
    normalise_row,
)

__all__ = ["StreamingRefreshSampler"]


class StreamingRefreshSampler:
    """Emit refresh cross-sections as arrivals complete them.

    Args:
        required_instruments: The names that must all refresh. At least two.
        max_staleness_ms: Age past which a complete row is flagged ``stale``.

    Notes:
        One sampler handles one partition at a time. Feeding it two partitions raises,
        rather than silently interleaving two unrelated sessions into one series.
    """

    def __init__(
        self, required_instruments: Iterable[Any], max_staleness_ms: int
    ) -> None:
        if isinstance(max_staleness_ms, bool) or not isinstance(max_staleness_ms, int):
            raise ValueError("max_staleness_ms must be a non-negative integer")
        if max_staleness_ms < 0:
            raise ValueError("max_staleness_ms must be a non-negative integer")

        instruments = _require_instruments(required_instruments)
        self._instruments = instruments
        self._allowed = set(instruments)
        self._max_staleness_ms = max_staleness_ms

        self._partition: str | None = None
        self._cursor = _BEFORE_TIME
        self._sequence = 0
        self._last_available_ms: int | None = None
        self._latest_by_record: dict[str, dict[str, Any]] = {}
        self._histories: dict[str, dict[str, Any]] = {}
        self._event_owners: dict[tuple[str, int], str] = {}
        self._pending: dict[str, list[dict[str, Any]]] = {
            instrument: [] for instrument in instruments
        }
        self._interval_arrivals: dict[str, int] = {
            instrument: 0 for instrument in instruments
        }
        self._interval_count = 0
        self._discarded_total = 0
        self._accepted = 0
        self._stale = 0
        self._input_updates = 0

    # --- ingestion ---------------------------------------------------------- #
    def observe(self, observation: Mapping[str, Any]) -> dict[str, Any] | None:
        """Ingest one arrival. Returns a refresh row if this arrival completed one."""

        # Per-row validation only. The cross-row revision rules are enforced by
        # _check_history below, which carries state the batch validator cannot see.
        row = normalise_row(observation, self._allowed)

        if self._partition is None:
            self._partition = row["partition"]
        elif row["partition"] != self._partition:
            raise ValueError(
                "one sampler handles one partition; "
                f"got {row['partition']!r} after {self._partition!r}"
            )

        if (
            self._last_available_ms is not None
            and row["available_ms"] < self._last_available_ms
        ):
            raise ValueError(
                "observations must arrive in non-decreasing available_at order"
            )
        self._check_history(row)

        self._last_available_ms = row["available_ms"]
        self._input_updates += 1
        self._interval_count += 1
        self._interval_arrivals[row["instrument"]] += 1
        self._latest_by_record[row["record_id"]] = row
        self._pending[row["instrument"]].append(row)

        if any(not queue for queue in self._pending.values()):
            return None
        return self._emit(row["available_ms"])

    def observe_many(
        self, observations: Iterable[Mapping[str, Any]]
    ) -> list[dict[str, Any]]:
        """Ingest many arrivals, returning every refresh row they completed."""

        emitted: list[dict[str, Any]] = []
        for observation in observations:
            row = self.observe(observation)
            if row is not None:
                emitted.append(row)
        return emitted

    def _check_history(self, row: dict[str, Any]) -> None:
        event_key = (row["instrument"], row["event_ms"])
        owner = self._event_owners.setdefault(event_key, row["record_id"])
        if owner != row["record_id"]:
            raise ValueError(
                "ambiguous duplicate event_time across distinct record_id values"
            )

        previous = self._histories.get(row["record_id"])
        if previous is None:
            if row["revision"] != 0:
                raise ValueError("the first revision of a record must be zero")
        else:
            if row["instrument"] != previous["instrument"]:
                raise ValueError("a correction cannot change instrument")
            if row["revision"] != previous["revision"] + 1:
                raise ValueError("record revisions must be contiguous")
            if row["available_ms"] <= previous["available_ms"]:
                raise ValueError("corrections must become available strictly later")
        self._histories[row["record_id"]] = row

    def _emit(self, refresh_ms: int) -> dict[str, Any]:
        selected = _select(self._latest_by_record, self._instruments)
        controllers = sorted(
            instrument
            for instrument, queue in self._pending.items()
            if queue[0]["available_ms"] == refresh_ms
        )
        kept = {
            (row["record_id"], row["revision"])
            for row in selected.values()
            if self._cursor < row["available_ms"] <= refresh_ms
        }
        discarded = self._interval_count - len(kept)
        self._discarded_total += discarded

        ages = {
            instrument: refresh_ms - row["event_ms"]
            for instrument, row in selected.items()
        }
        status = "stale" if max(ages.values()) > self._max_staleness_ms else "accepted"
        if status == "stale":
            self._stale += 1
        else:
            self._accepted += 1
        self._sequence += 1

        emitted = {
            "partition": self._partition,
            "sequence": self._sequence,
            "previous_refresh_available_at": (
                None if self._cursor == _BEFORE_TIME else format_timestamp_ms(self._cursor)
            ),
            "refresh_available_at": format_timestamp_ms(refresh_ms),
            "controller_instruments": controllers,
            "status": status,
            "values": {
                instrument: selected[instrument]["value"]
                for instrument in self._instruments
            },
            "sources": {
                instrument: {
                    "record_id": selected[instrument]["record_id"],
                    "revision": selected[instrument]["revision"],
                    "event_time": selected[instrument]["event_time"],
                    "available_at": selected[instrument]["available_at"],
                }
                for instrument in self._instruments
            },
            "event_age_ms": ages,
            "arrivals_by_instrument": dict(sorted(self._interval_arrivals.items())),
            "discarded_updates": discarded,
        }

        self._cursor = refresh_ms
        # Every queue is non-empty here, so dropping one entry each keeps exactly the
        # arrivals that have not yet been consumed by a refresh.
        for instrument in self._instruments:
            self._pending[instrument] = [
                row
                for row in self._pending[instrument]
                if row["available_ms"] > refresh_ms
            ]
        self._interval_count = sum(len(queue) for queue in self._pending.values())
        self._interval_arrivals = {
            instrument: len(self._pending[instrument])
            for instrument in self._instruments
        }
        return emitted

    # --- introspection ------------------------------------------------------ #
    @property
    def summary(self) -> dict[str, Any]:
        """The same partition summary the batch function reports."""

        tail = sum(len(queue) for queue in self._pending.values())
        return {
            "partition": self._partition,
            "input_updates": self._input_updates,
            "refresh_candidates": self._sequence,
            "accepted_rows": self._accepted,
            "stale_rows": self._stale,
            "discarded_updates": self._discarded_total,
            "unmatched_tail_updates": tail,
            "loss_fraction": {
                "numerator": self._discarded_total + tail,
                "denominator": self._input_updates,
            },
        }

    @property
    def waiting_for(self) -> list[str]:
        """Instruments that have not posted since the last refresh.

        This is the live version of the controller question: whatever is in this list
        is what the whole basket is currently waiting on.
        """

        return sorted(
            instrument for instrument, queue in self._pending.items() if not queue
        )

    @property
    def refresh_count(self) -> int:
        return self._sequence

    @property
    def last_refresh_at(self) -> str | None:
        return None if self._cursor == _BEFORE_TIME else format_timestamp_ms(self._cursor)
